src/main/java/io/airlift/units/Count.java

```diff
@@ -0,0 +1,205 @@
+
+    public Count(long count, Magnitude magnitude)
+    {
+        checkArgument(!Double.isInfinite(count), "count is infinite");
+        checkArgument(!Double.isNaN(count), "count is not a number");
+        checkArgument(count >= 0, "count is negative");
+        requireNonNull(magnitude, "magnitude is null");
+
+        this.value = count;
+        this.magnitude = magnitude;
+    }
+
+    public long getValue()
+    {
+        return value;
+    }
+
+    public Magnitude getMagnitude()
+    {
+        return magnitude;
+    }
+
+    public long getValue(Magnitude magnitude)
+    {
+        if (value == 0L) {
+            return 0L;
+        }
+
+        long scale = this.magnitude.getFactor() / magnitude.getFactor();
+        if (scale == 0) {
+            throw new IllegalArgumentException(format("Unable to return value %s in magnitude %s due precision loss", this, magnitude));
+        }
+        return value * scale;
+    }
+
+    public Count convertTo(Magnitude magnitude)
+    {
+        requireNonNull(magnitude, "magnitude is null");
+        return new Count(getValue(magnitude), magnitude);
+    }
+
+    /**
+     * @return converted count with no bigger value than 1000 in succinct magnitude, fractional part is rounded
+     */
+    public Count convertToMostSuccinctRounded()
+    {
+        for (Magnitude magnitude : MAGNITUDES) {
+            double converted = (double) value * this.magnitude.getFactor() / magnitude.getFactor();
+            if (converted < 1000) {
+                return new Count(Math.round(converted), magnitude);
+            }
+        }
+        throw new IllegalStateException();
+    }
+
+    @JsonValue
+    @Override
+    public String toString()
+    {
+        //noinspection FloatingPointEquality
+        if (floor(value) == value) {
+            return (long) (floor(value)) + magnitude.getMagnitutdeString();
+        }
+
+        return format(Locale.ENGLISH, "%.2f%s", value, magnitude.getMagnitutdeString());
+    }
+
+    @JsonCreator
+    public static Count valueOf(String count)
+            throws IllegalArgumentException
+    {
+        requireNonNull(count, "count is null");
+        checkArgument(!count.isEmpty(), "count is empty");
+
+        Matcher matcher = PATTERN.matcher(count);
+        if (!matcher.matches()) {
+            throw new IllegalArgumentException("Not a valid count string: " + count);
+        }
+
+        long value = parseLong(matcher.group(1));
+        String magnitutdeString = matcher.group(2);
+
+        for (Magnitude magnitude : Magnitude.values()) {
+            if (magnitude.getMagnitutdeString().equals(magnitutdeString)) {
+                return new Count(value, magnitude);
+            }
+        }
+
+        throw new IllegalArgumentException("Unknown magnitude: " + magnitutdeString);
+    }
+
+    @Override
+    public int compareTo(Count o)
+    {
+        return Double.compare(getValue(Magnitude.SINGLE), o.getValue(Magnitude.SINGLE));
```

Comment:
This double comparison is wrong. However, using `getValue()` won't work, since it can fail on overflow. Maybe we should have a `toBigInteger()` method and use that here.